src/Build/Instance/ProjectInstance.cs

```diff
@@ -1314,11 +1329,28 @@ ICollection<ProjectItemInstance> IItemProvider<ProjectItemInstance>.GetItems(str
         /// immutable if we are immutable.
         /// Only called during evaluation, so does not check for immutability.
         /// </summary>
-        ProjectPropertyInstance IEvaluatorData<ProjectPropertyInstance, ProjectItemInstance, ProjectMetadataInstance, ProjectItemDefinitionInstance>.SetProperty(string name, string evaluatedValueEscaped, bool isGlobalProperty, bool mayBeReserved)
+        ProjectPropertyInstance IEvaluatorData<ProjectPropertyInstance, ProjectItemInstance, ProjectMetadataInstance, ProjectItemDefinitionInstance>.SetProperty(string name, string evaluatedValueEscaped, bool isGlobalProperty, bool mayBeReserved, bool isEnvironmentVariable)
         {
             // Mutability not verified as this is being populated during evaluation
             ProjectPropertyInstance property = ProjectPropertyInstance.Create(name, evaluatedValueEscaped, mayBeReserved, _isImmutable);
             _properties.Set(property);
+
+            if (_trackPropertyReads)
+            {
+                if (isEnvironmentVariable)
+                {
+                    // If it's an env var, track it.
+                    _environmentVariables.Add(name);
+                }
+                else if (_environmentVariables.Contains(name))
```

Comment:
You still have some cases of "if contains, then remove" or "if not contains, then add" in the context of `HashSet<T>`. In both cases, you can remove the contains to halve the lookups. The collection does not throw if you add an entry that already exists, nor if you remove an entry that does not exist.

Reply:
Removed with latest update.